### PR TITLE
fix: bound Codex cost scans on giant session corpora

### PR DESCRIPTION
**Pull request by D4ilyHub**
## Summary

- Bound Codex local token-cost scans so giant rollout corpora cannot monopolize a CPU core for hours.
- Process newest rollout files first and cap newly read bytes per refresh.
- Resume oversized files from persisted per-file progress until every complete JSONL record is covered; no rollout is permanently omitted from cost history.
- Keep account and rate-limit probing unchanged. This work only affects local session cost history.

## Resumable parsing

The cache now records the rollout file identity, target size, byte offset, reducer state, and any partial JSONL record needed to continue safely. Each refresh consumes at most the configured per-file slice and remaining global byte budget. A changed identity or size invalidates incompatible progress instead of mixing histories.

Ordinary rollouts publish accumulated complete records as they advance. Subagent rollouts retain compact relevant events until the final chunk establishes their copied-prefix ownership boundary, then publish the same result as an unbounded parse. Completed files follow the existing warm-cache path.

## Contributor evidence

@D4ilyHub diagnosed the original scan storm and validated the bounded/newest-first machinery against a real 276 GiB corpus containing 1,322 rollouts. Before the change, the scanner could pin a core for hours and immediately resume after relaunch. The contributor's cold proof completed in about 16 seconds wall time and 6 seconds CPU under a 512 MiB refresh budget, while account and rate-limit display continued independently.

The maintainer follow-up replaces the initial permanent 256 MiB file skip with resumable bounded parsing while preserving that bounded-refresh behavior.

## Verification

- Focused scanner suites cover bounded multi-refresh convergence, cache round-trip persistence, a single JSONL record larger than the slice, newest-first ordering, budget admission, JSONL edge cases, and compact subagent accounting.
- Local scanner proof: 34 tests passed across the performance, JSONL, and compact-subagent suites.
- Repository format and lint checks pass.
- Hosted CI is the authoritative full-suite proof for this PR.

### CODE REVIEW

**Comment by D4ilyHub** [comment]
## Follow-up: addressed Codex P1 review comments

Pushed `49d09668` on this branch.

### P1 — forced full rescans not charged
- **Fixed.** `pendingCodexScanWorkBytes` no longer returns 0 for size/mtime-matched cache entries after `keepCached` rejects them.
- Forced rescans now charge full file bytes against per-file and per-refresh budgets.

### P1 — fork parent snapshot reads unbounded
- **Fixed.** `CodexInheritedTotalsResolver` uses the shared scan budget.
- Oversized/deferred parents skip the full baseline parse and surface as unresolved (existing fork handling).
- Successful parent reads consume budget.

### Tests added
- forced-rescan work sizing for unchanged multi-GB cache entries
- small fork child + oversized parent completes quickly without thrashing

### Local verification
- `swift build --target CodexBarCore` passes on this machine
- Full `CostUsagePerformanceGateTests` still needs Xcode CI here (CLT-only host)

Thanks for the sharp catches — both were real bypasses of the original protection.

File: Sources/CodexBarCore/Vendored/CostUsage/CostUsageScanner.swift [hunk 1]
```diff
@@ -2732,10 +2788,60 @@ enum CostUsageScanner {
...
             return
         }
         try Self.rescanCodexFile(input: input, context: context, cache: &cache, state: &state)
+        context.scanBudget?.consume(workBytes: pendingWorkBytes)
+    }
+
+    static func pendingCodexScanWorkBytes(metadata: CodexFileMetadata, cached: CostUsageFileUsage?) -> Int64 {
+        guard let cached else { return max(0, metadata.size) }
+        if cached.mtimeUnixMs == metadata.mtimeUnixMs, cached.size == metadata.size {
+            return 0
```
[the remark this answers is not in the document]
Comment (D4ilyHub) [reply Sources/CodexBarCore/Vendored/CostUsage/CostUsageScanner.swift]: Agreed — valid hole. Fixed in 49d09668.

`pendingCodexScanWorkBytes` is only used after `keepCachedCodexFileIfFresh` fails. The old size/mtime-equality branch could still report `0` for forced full rescans (window/pricing/project-metadata/priority invalidation), so the budget admitted multi-GB files and then charged nothing after `rescanCodexFile`.

That zero branch is gone. Unchanged-but-must-rescan files now charge full file size; only true grow+non-fork incremental tails charge the delta.

File: Sources/CodexBarCore/Vendored/CostUsage/CostUsageScanner.swift [hunk 2]
```diff
@@ -2732,10 +2788,60 @@ enum CostUsageScanner {
         if try Self.keepCachedCodexFileIfFresh(input: input, context: context, cache: &cache, state: &state) {
             return
         }
+
+        let pendingWorkBytes = Self.pendingCodexScanWorkBytes(metadata: metadata, cached: cached)
+        if let budget = context.scanBudget {
+            switch budget.admit(workBytes: pendingWorkBytes) {
```
[the remark this answers is not in the document]
Comment (D4ilyHub) [reply Sources/CodexBarCore/Vendored/CostUsage/CostUsageScanner.swift:3142]: Agreed — also valid for fork/subagent corpora. Fixed in 49d09668.

`CodexInheritedTotalsResolver` now shares the same `CodexScanBudget`. Before `parseCodexTokenSnapshots` on a parent, we admit the parent file size; oversized/deferred parents return an unresolved baseline (existing fork path) instead of a full multi-GB dependency parse. Successful parent reads consume budget so they count toward the per-refresh cap.

Added a performance gate covering small-child + oversized-parent.

**Comment by D4ilyHub** [comment]
## Real behavior proof added (ClawSweeper request)

Updated the PR body with redacted after-fix live runtime evidence on head `49d096683836`.

### What was run
Public `CostUsageFetcher` against the **actual large local `~/.codex` session corpus** (not a toy fixture), plus the synthetic 300 MiB oversized-sibling integration.

### Corpus (aggregate / redacted)
- **276.0 GiB** sessions tree
- **1322** `rollout-*.jsonl` files
- **417** files ≥256 MiB (3 ≥1 GiB)
- Largest sizes (MiB only): 1335, 1308, 1294, 722, 715, …

### After-fix cold catch-up (forceRefresh, historyDays=30, empty temp cache)
```
elapsed_s=16.073
process_cpu_s=5.955
bounded=yes
```

CodexBarCore work-limit log (path-free):
```
skippedOversized=424
deferredByBudget=866
bytesConsumed=536854072   # ~512 MiB refresh budget
maxFileBytes=268435456
maxBytesPerRefresh=536870912
```

### Second refresh (same cache)
```
elapsed_s=6.538
still_bounded=yes
total_tokens increased (progressive catch-up under budget)
```

### Synthetic oversized sibling
```
300 MiB giant skipped; small session totalTokens=110; elapsed=0.021s
```

### Intentional tradeoff (explicit for merge review)
Local token-cost history **can lag** for deferred/oversized multi-hundred-MB and multi-GB rollouts. That is the availability tradeoff. Account/rate-limit remaining + reset remain on a **separate path** and are unaffected by these scan budgets.

### Not runnable on author host
Full `CostUsagePerformanceGateTests` still needs Xcode/Swift Testing (CLT-only here). Gates are in-tree for CI.

@clawsweeper re-review

**Comment by steipete** [comment]
Merged as `53376361a83e619204fa4e08c85ec64c5dfe4fb6`.

Thank you @D4ilyHub for the 276 GiB live evidence. That proof made the ownership requirement clear: the cost scanner itself—not callers or a permanent skip list—must own the per-file/per-refresh budget and persist resumable JSONL progress. Giant rollout files now stay bounded on each refresh and continue across later refreshes, including parent/subagent baseline work, so no file is permanently omitted.
